### PR TITLE
ENH: add skeleton of interactive plan creation

### PR DESCRIPTION
**Pull request by tacaswell**
- Adds a base-class to handle the mpl related work
 - Add one example of specializing to an outer_product_plan

attn @licode

### CODE REVIEW

File: bluesky/interactive.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        'If the widget should be active'
+        return self.widget.get_active()
+
+    @active.setter
+    def active(self, active):
+        return self.widget.set_active(active)
+
+
+class OuterProductWidget(ROIPlanCreator):
+    def __init__(self, ax, dets, m1, m2, numsteps1, numsteps2=None,
```
Comment (klauer) [inline bluesky/interactive.py]: Why don't these arg names match what bluesky docstrings call them?

File: bluesky/interactive.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+    def _plan(self, p1, p2):
+        x1, y1 = p1
+        x2, y2 = p2
+
+        local_md = {'interactive': {
+            'created_by': self.__class__.__name__,
+            'rect_points': {'p1': p1, 'p2': p2}}}
+
+        md = ChainMap(self.md, local_md)
```
Comment (licode) [inline bluesky/interactive.py]: what's the difference between self.md and local_md?

File: bluesky/interactive.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        '''data coordinates of the first and second corners.
+
+        Coordinates are lower left and upper right.
+
+        Returns
+        -------
+        points : tuple
+           ((x1, y1), (x2, y2))
+        '''
+        # TODO deal with ordering issues
```
Comment (licode) [inline bluesky/interactive.py]: Can you elaborate more on ordering issues?

**Comment by licode** [comment]
examples or tests on this function?
Just left some simple questions. 
LGTM 👍

File: bluesky/interactive.py (hunk 2 above)
Comment (tacaswell) [inline bluesky/interactive.py]: `self.md` hold the metadata that the user passed in on init, `local_md` is metadata we generated as part of this call.

File: bluesky/interactive.py (hunk 3 above)
Comment (tacaswell) [inline bluesky/interactive.py]: that comment should not have survived.  I was not sure if mpl return the points in the order you drew them or (bottom left, top right) and it turns out it always is (BL, UR).

File: bluesky/interactive.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        x1, y1 = p1
+        x2, y2 = p2
+
+        local_md = {'interactive': {
+            'created_by': self.__class__.__name__,
+            'rect_points': {'p1': p1, 'p2': p2}}}
+
+        md = ChainMap(self.md, local_md)
+
+        return outer_product_scan(self.dets,
```
Comment (klauer) [inline bluesky/interactive.py]: Having access to `plan_args` before creating the plan would be nice, too.

File: bluesky/interactive.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        if md is None:
+            md = {}
+        self.md = md
+
+    def _plan(self, p1, p2):
+        x1, y1 = p1
+        x2, y2 = p2
+
+        local_md = {'interactive': {
+            'created_by': self.__class__.__name__,
```
Comment (klauer) [inline bluesky/interactive.py]: `created_by_widget`?

File: bluesky/interactive.py (hunk 4 above)
Comment (tacaswell) [inline bluesky/interactive.py]: Not sure what you mean, the `plan_args` should be put into the md by the underlying plan.

Comment (klauer) [inline bluesky/interactive.py]: I mean `widget.plan_args` would give back that information without creating a plan.

Comment (tacaswell) [inline bluesky/interactive.py]: What is the use case for that?  I would rather keep the 'required' implementation as simple as possible (`__init__` + `_plan`).

Comment (klauer) [inline bluesky/interactive.py]: Re-use / tweak without rewriting - pretty much what I'm always pushing for

**Comment by tacaswell** [comment]
I rearranged the meta-data capture to make `_plan` extremely thin.

File: bluesky/interactive.py (hunk 3 above)
Comment (danielballan) [inline bluesky/interactive.py]: Can you delete that comment, then?

**Comment by danielballan** [comment]
Rather than spin Travis again, I'll just merge as is and cleanup that potentially-confusing comment separately....

**Pull request merged**